Msi 2286 in store pickup checkout implementation on weabpi level

swnsma opened the pull request:
### Fixed Issues (if relevant)
1. magento-engcom/msi#2286

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Add base module declaration.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Add DB Schema declaration.

swnsma committed:
Merge branch 'store-pickup' into MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Add save and load of pickup location exetension attribute for Quote Address entity.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Improve code consistency.

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/Plugin/Quote/Address/LoadPickupLocationForQuoteAddress.php:
```diff
@@ -0,0 +1,69 @@
+        $this->addressExtensionInterfaceFactory = $addressExtensionInterfaceFactory;
+    }
+
+    /**
```
Please add comment block here

swnsma replied:
Done.

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/Plugin/Quote/Address/LoadPickupLocationForQuoteAddress.php:
```diff
@@ -0,0 +1,69 @@
+        }
+
+        $pickupLocationCode = $this->getPickupLocationCodeByQuoteAddressId->execute((int)$entity->getId());
+        if ($pickupLocationCode) {
```
Please process the negative case first to avoid nested **if** operators and reduce code complexity.

swnsma replied:
Done.

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/Plugin/Quote/Address/SavePickupLocationForQuoteAddress.php:
```diff
@@ -0,0 +1,53 @@
+        $this->saveQuoteAddressPickupLocation = $saveQuoteAddressPickupLocation;
+    }
+
+    /**
```
Please add a comment here

swnsma replied:
Done.

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/Plugin/Quote/AddressCollection/GetPickupLocationInformationPlugin.php:
```diff
@@ -0,0 +1,90 @@
+        $this->addressExtensionInterfaceFactory = $addressExtensionInterfaceFactory;
+    }
+
+    /**
```
Please add comment here

swnsma replied:
Done.

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/Plugin/Quote/AddressCollection/GetPickupLocationInformationPlugin.php:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * Add Loaded Pickup Location to Extension Attributes.
+     *
+     * @param Address $item
```
Please add return type here to doc block

swnsma replied:
Done.

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/etc/di.xml:
```diff
@@ -0,0 +1,16 @@
+        <plugin name="save_pickup_location_for_quote_address" type="Magento\InventoryInStorePickupQuote\Plugin\Quote\Address\SavePickupLocationForQuoteAddress" sortOrder="20" />
+        <plugin name="load_pickup_location_for_quote_address" type="Magento\InventoryInStorePickupQuote\Plugin\Quote\Address\LoadPickupLocationForQuoteAddress" sortOrder="30" />
+    </type>
+</config>
```
Please add the new line at the end of file

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/etc/extension_attributes.xml:
```diff
@@ -0,0 +1,12 @@
+    <extension_attributes for="Magento\Quote\Api\Data\AddressInterface">
+        <attribute code="pickup_location_code" type="string" />
+    </extension_attributes>
+</config>
```
Please add the new line at the end of file

swnsma replied:
It already in place)
Github do not show empty line :)

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/etc/module.xml:
```diff
@@ -0,0 +1,15 @@
+            <module name="Magento_InventoryInStorePickupApi" />
+        </sequence>
+    </module>
+</config>
```
Please add the new line at the end of file

swnsma replied:
Already in place.

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/registration.php:
```diff
@@ -0,0 +1,12 @@
+    \Magento\Framework\Component\ComponentRegistrar::MODULE,
+    'Magento_InventoryInStorePickupQuote',
+    __DIR__
+);
```
Could I ask you to add the new line at the end of file everywhere:))

swnsma replied:
Already in place.

Stepa4man reviewed (changes requested):
Hi @swnsma, Thank you for your contribution!
Could I ask you to check my comments?

ishakhsuvarov commented:
@swnsma We might consider moving Checkout related APIs to `InStorePickupStorefrontApi` module

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Add code review comments.

Commit:
Merge branch 'store-pickup' into MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level

Commit:
Merge branch 'store-pickup' into MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Add service to provide Pickup Location by Source Code, Sales Channel and Sales Channel Type.

Stepa4man commented on app/code/Magento/InventoryInStorePickupApi/Api/GetPickupLocationInterface.php:
```diff
@@ -0,0 +1,34 @@
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
```
Please remove declaring strict types from php interfaces
see 1.3.1 https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html#1-basic-programming-principles

swnsma replied:
Good point to discuss.
@ishakhsuvarov why we generally not follow this recommendation?

ishakhsuvarov replied:
@swnsma It became a standard in MSI repository before 1.3.1 was introduced. I guess this is a power of habit at this point. Personally I don't see any harm in it, but of course we would like to follow guidelines for new code.

Stepa4man reviewed (changes requested):
@swnsma, one more comment. Please take a look.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Fix phantom issue with missed form element in Source Edit form.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Add public constant for In-Store Pickup Delivery Method.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Replace Pickup Location address for Shipping address during checkout flow.

swnsma commented:
TODO:
- add Pickup Location registry
- copy Pickup Location code to created order
- discuss points

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Add TODO for future code simplification.
Add possibility to place In-Store Pickup order.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Add dependencies.
Remove 'declare(strict_types=1);' from interface.
Switch save of Quote Address Pickup Location from Resource Model after save to Address Entity after AfterSave to decrease number of writes to DB.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Adjust tests.

Commit:
Merge branch 'store-pickup' into MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level

swnsma commented:
> add Pickup Location registry

Skipped. Endpoint to receive Pickup Location will be refactored.

> copy Pickup Location code to created order

Done with a Plugin. Need to be refactored when core issue will be solved: [magento/magento2#23386](https://github.com/magento/magento2/issues/23386)

@ishakhsuvarov, @Stepa4man, gentlemens, please do code review.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Fix failed tests.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Remove dependency from Magento Checkout.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Remove excess code.

ishakhsuvarov commented:
@magento run all tests

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/Model/ResourceModel/GetPickupLocationCodeByQuoteAddressId.php:
```diff
@@ -0,0 +1,55 @@
+        $connection = $this->connection->getConnection();
+        $table = $this->connection->getTableName('inventory_pickup_location_quote_address');
+
+        $select = $connection->select()
```
Why don't you use Magento DB abstraction? I don't see anything special here to use directly Zend Select?

swnsma replied:
I am not sure if I understand your comment. 
`$connection->select()` returns `\Magento\Framework\DB\Select`.

swnsma replied:
Discussed: to skip.

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/Plugin/Quote/ReplaceShippingAddressForShippingAddressManagement.php:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    /**
+     * @param ShippingAddressManagementInterface $subject
```
Please add comment here

swnsma replied:
Added.

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/etc/module.xml:
```diff
@@ -0,0 +1,18 @@
+ */
+-->
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
+    <module name="Magento_InventoryInStorePickupQuote" setup_version="1.0.0">
```
Please get rid of setup version, it's not used any more

swnsma replied:
Good to know that this is possible :)
Done.

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/Plugin/Quote/ReplaceShippingAddressForShippingAddressManagement.php:
```diff
@@ -0,0 +1,142 @@
+        ShippingAddressManagementInterface $subject,
+        int $cartId,
+        AddressInterface $address
+    ): array {
```
The method is pretty overloaded, don't think? Please consider to move logic to the service and divide it into the logical parts.

swnsma replied:
Any suggestions?

Stepa4man replied:
We can set up the call to discuss this.

swnsma replied:
Discussed: refactored.
Please review.

Stepa4man reviewed (changes requested):
Hi @swnsma, please check out more comment. Thanks in advance!

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Apply code review changes.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Refactoring to decrease number of dupplicate code.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Refactoring.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Code reformating.

swnsma commented:
Special thanks to @Stepa4man who pushed me to refactor this peace of legacy-depended code.
Could you continue review?

swnsma commented:
Case to check: 
- create quote
- add product
- set shipping information with shipping address, billing address and In-Store Pickup delivery method
- add payment infromation
- place the order

First time you will try to place order, error will be occurred.
Probably the issue can be related to \Magento\Quote\Model\Quote\Address::_populateBeforeSaveData

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Fix static tests.

Stepa4man commented on app/code/Magento/InventoryInStorePickupQuote/Plugin/Quote/PassPickupLocationToOrder.php:
```diff
@@ -31,7 +31,8 @@ public function beforeConvert(ToOrder $subject, Address $address, array $data =
 
         if ($extension && $extension->getPickupLocationCode()) {
-            $data['extension_attribute_pickup_location_code_pickup_location_code'] = $extension->getPickupLocationCode();
+            $data['extension_attribute_pickup_location_code_pickup_location_code'] =
```
Please consider moving this string to constant. Not sure if this constant should be defined in this class. How do you think?

swnsma replied:
Hopefully, after merge 2.3-develop from magento/magento2, information about this field will be removed.
I have moved it to constants in 
\Magento\InventoryInStorePickupQuote\Plugin\Quote\PassPickupLocationToOrder
\Magento\InventoryInStorePickupQuote\Plugin\Quote\ReplaceShippingAddressForShippingAddressManagement

In our case there are fields for different entities which have the same name.

Stepa4man commented on app/code/Magento/InventoryInStorePickupShipping/Model/IsInStorePickupDeliveryCart.php:
```diff
@@ -0,0 +1,38 @@
+
+        return $shipping->getMethod() === InStorePickup::DELIVERY_METHOD;
+    }
+}
```
Please add new line at the end of file

swnsma replied:
Fixed.

Stepa4man reviewed (dismissed):
@swnsma, Thanks for the patience. Please check my comments.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Refactoring.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Refactoring.
Verify Quote Address by Address Shipping Method instead of Quote Shipping Assignment.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Fix integration tests.

ishakhsuvarov committed:
Merge branch 'store-pickup' of github.com:magento-engcom/msi into MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickupQuote/Model/GetWebsiteCodeByStoreId.php:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Provide website code by store id.
+ */
+class GetWebsiteCodeByStoreId
```
I'm not sure if Store Pickup module is a good location for this kind of service.

swnsma replied:
I agree. Does it real option just to mark it with '@internal'?

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickupQuote/Model/ResourceModel/SaveQuoteAddressPickupLocation.php:
```diff
@@ -0,0 +1,54 @@
+     */
+    public function execute(int $addressId, string $pickupLocationCode): void
+    {
+        $connection = $this->connection->getConnection();
```
Is it possible to just use existing quote address tables to save pickup location address?

swnsma replied:
Do you mean: add 'pickup_location' as an attribute to Quote Address entity and store it into the 'quote_address' table?

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickupQuote/Model/ToQuoteAddress.php:
```diff
@@ -0,0 +1,96 @@
+            AddressInterface::class
+        );
+
+        $address->setShippingMethod(InStorePickup::DELIVERY_METHOD);
```
Why shipping method is a responsibility of this class?

ishakhsuvarov committed:
Static tests stabilization

ishakhsuvarov committed:
Skip invalid integration test

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Simplify fixtures.
Fix failed integration test.
Remove Pickup Location shipping address assignment from address copy service.

Commit:
Merge branch 'MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level' of github.com:magento-engcom/msi into MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Do not skip the test.

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Fix simplified name.

ishakhsuvarov committed:
Static tests

Commit:
MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level.
Adjust test.

Commit:
Merge branch 'MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level' of github.com:magento-engcom/msi into MSI-2286-In-Store-Pickup-checkout-implementation-on-Weabpi-level

ishakhsuvarov commented:
Semantic Version Checker failure is expected as new APIs are added into Store Pickup.

swnsma commented:
<3
